proposals/0184-block-writeable-account-data-limit.md

```diff
@@ -0,0 +1,107 @@
+---
+simd: '0184'
+title: Block Writeable Account Data Limit
+authors:
+  - Brennan Watt (Anza)
+category: Standard
+type: Core
+status: Idea
+created: 2024-10-15
+feature: (fill in with feature tracking issues once accepted)
+supersedes:
+superseded-by:
+extends:
+---
+
+## Summary
+
+A protocol limit on the amount of account data in bytes that may be written
+during a single block.
+
+## Motivation
+
+When direct mapping is enabled, clients will be able to handle much more account
+data because there will be fewer copies in memory (e.g. for each VM invocation /
+CPI call). However, writeable account data cannot be direct mapped because it
+needs to be written back after a successful tx execution. This necessitates a
+decoupling between account read and write data in order to increase readable
+account data limits.
+
+Note: There is an existing per block limit on newly allocated account data
+(100MB), but this does not encompass account data being updated/written back.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+This feature would check the amount of account data marked as writeable for each
+transaction being handled by the block producer (for inclusion in a block) or
+validator (replaying transactions in a block), add it to a running total of
+total writeable account data for the block, and compare it against some
+threshold (recommend 2GB to start) to decide if the transaction is allowed to be
+included in the block (if sum is below the limit for the case of block producer)
+or if the slot needs to be marked dead (if sum is above the limit for the case
+of validator).
+
+This feature to limit writeable account data per block would slot in nicely to
+the existing cost model and cost tracker that maintain per block limits for
+various resources transactions can consume. This will allow block producers to
+determine when a block is "full" and validators to mark slots dead that do not
+adhere to these limits.
+
+This feature could be implemented like so (using Agave client as reference):
+
+1. Update the cost model to compute the aggregate size of all accounts marked
+  writeable for each transaction. This can be done similarly to how
+  `data_bytes_len_total` is computed inside `get_transaction_cost` but limited
+  to accounts marked writeable.
```

Comment:
I think this is inaccurate.

Looking at the `get_transaction_cost()` in `cost-model`, I think, `data_bytes_len_total` only calculates the size of the instruction data passed into instructions.
It does not look at the account sizes:
```rust
            data_bytes_len_total =
                data_bytes_len_total.saturating_add(instruction.data.len() as u64);
```
https://github.com/anza-xyz/agave/blob/f9f8b60ca15fa721c6cdd816c99dfd4e9123fd77/cost-model/src/cost_model.rs#L187-L188

The loaded accounts data size is only included in the transaction cost after the transaction execution.
Actual value is passed into the `calculate_cost_for_executed_transaction()` as the `actual_loaded_accounts_data_size_bytes` argument.
And it is populated and checked in the `blockstore_processor.rs`:
```rust
                Some(CostModel::calculate_cost_for_executed_transaction(
                    tx,
                    committed_tx.executed_units,
                    committed_tx.loaded_account_stats.loaded_accounts_data_size,
                    &bank.feature_set,
                ))
```
https://github.com/anza-xyz/agave/blob/f9f8b60ca15fa721c6cdd816c99dfd4e9123fd77/ledger/src/blockstore_processor.rs#L246-L251

We could record have a similar counter for the amount of writable data that the transaction touched.

So at the high level it can be computed similarly to what we are already doing.
It is just that the details would be different.

Not sure if there are any implications for the DX, due to the fact that we can only compute this value after a transaction execution.

Reply:
Yep, you're right. We don't have any precedent for understanding actual account data before loading/executing transactions. We rely on either scraping top level instructions (e.g. for allocate) or estimating and then adjusting. 

This would seemingly have to fall into the estimate + adjust case (unless we want to do something radical like grab account data size before execution), which would likely impact DX if we push this onto the user similar to how we ask for CU/Data budget estimates.

One option is to leverage the existing Data budget value for writes as well (defaults to 2MiB per tx or can be set by user). I think this is likely small enough of a default to not bottleneck scheduling before the data budget gets refunded post execution. We would need to study this.